streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala

```diff
@@ -162,6 +164,59 @@ class KTable[K, V](val inner: KTableJ[K, V]) {
+    * In order to assign a state, the state must be created and registered
+    * beforehand via stores added via `addStateStore` or `addGlobalStore` before they can be connected to the `Transformer`
+    *
+    * @param valueTransformerWithKeySupplier a instance of `ValueTransformerWithKeySupplier` that generates a `ValueTransformerWithKey`.
+    *                                 At least one transformer instance will be created per streaming task.
+    *                                 Transformer implementations doe not need to be thread-safe.
+    * @param stateStoreNames          the names of the state stores used by the processor
+    * @return a [[KStream]] that contains records with unmodified key and new values (possibly of different type)
+    * @see `org.apache.kafka.streams.kstream.KStream#transformValues`
+    */
+  def transformValues[VR](valueTransformerWithKeySupplier: ValueTransformerWithKeySupplier[K, V, VR],
+                          stateStoreNames: String*): KTable[K, VR] = {
+    inner.transformValues[VR](valueTransformerWithKeySupplier, stateStoreNames: _*)
+  }
+
+  /**
+    * Create a new `KTable` by transforming the value of each record in this `KTable` into a new value, (with possibly new type).
+    * A `ValueTransformer` (provided by the given `ValueTransformerSupplier`) is applied to each input
+    * record value and computes a new value for it.
+    * This is similar to `#mapValues(ValueMapperWithKey)`, but more flexible, allowing stateful, rather than stateless,
+    * record-by-record operation, access to additional state-stores, and access to the `ProcessorContext`.
+    * In order to assign a state, the state must be created and registered
+    * beforehand via stores added via `addStateStore` or `addGlobalStore` before they can be connected to the `Transformer`
+    * The resulting `KTable` is materialized into another state store (additional to the provided state store names)
+    * as specified by the user via `Materialized` parameter, and is queryable through its given name.
+    *
+    * @param valueTransformerWithKeySupplier a instance of `ValueTransformerWithKeySupplier` that generates a `ValueTransformerWithKey`
+    *                                 At least one transformer instance will be created per streaming task.
+    *                                 Transformer implementations doe not need to be thread-safe.
+    * @param materialized             an instance of `Materialized` used to describe how the state store of the
+    *                                 resulting table should be materialized.
+    * @param stateStoreNames          the names of the state stores used by the processor
+    * @return a [[KStream]] that contains records with unmodified key and new values (possibly of different type)
+    * @see `org.apache.kafka.streams.kstream.KStream#transformValues`
+    */
+  def transformValues[VR](valueTransformerWithKeySupplier: ValueTransformerWithKeySupplier[K, V, VR],
```

Comment:
There are some proposals on refactoring the scala api to make better leverage on its type inference: https://github.com/apache/kafka/pull/5019

I think to be consistent with that PR the API should be 

```
transformValues[VR](valueTransformerWithKeySupplier: ValueTransformerWithKeySupplier[K, V, VR]) (materialized: Materialized[K, VR, KeyValueStore[Bytes, Array[Byte]]],
                          stateStoreNames: String*): KTable[K, VR])
```

I will let @joan38 to chime in here as well:

Reply:
It's been a while since i've looked at Scala, can I suggest such changes are left to the linked PR? i.e. we try and merge this one first, as there seems to be open questions of the other PR.

Reply:
It should be ok here since `valueTransformerWithKeySupplier` doesn't have parameters that depends on another.